contracts/PrepaidCardMarket_v2.sol

```diff
@@ -0,0 +1,304 @@
+  address public tokenManager;
+  address public provisioner;
+  address public versionManager;
+  address public exchangeAddress;
+  mapping(address => mapping(address => uint256)) public balance; // issuer safe address -> token -> balance
+  mapping(address => address) public issuers; // issuer safe address -> issuer EOA
+  mapping(bytes32 => SKU) public skus; // sku => sku data
+  mapping(bytes32 => uint256) public asks; // sku => ask price (in issuing token)
+
+  bool public paused;
+
+  EnumerableSetUpgradeable.AddressSet internal trustedProvisioners;
+
+  event Setup();
+
+  event TokensDeposited(
+    address issuer,
+    uint256 amount,
+    address token,
+    address safe
+  );
+  event TokensWithdrawn(
+    address safe,
+    address issuer,
+    address token,
+    uint256 amount
+  );
+  event SkuAdded(
+    address issuer,
+    address issuingToken,
+    uint256 faceValue,
+    string customizationDID,
+    bytes32 sku
+  );
+  event AskSet(
+    address issuer,
+    address issuingToken,
+    bytes32 sku,
+    uint256 askPrice
+  );
+  event ProvisionedPrepaidCard(
+    address prepaidCard,
+    address customer,
+    bytes32 sku,
+    uint256 askPrice
+  );
+  event TrustedProvisionerAdded(address token);
+  event TrustedProvisionerRemoved(address token);
+  event PausedToggled(bool paused);
+
+  function initialize(address owner) public override initializer {
+    paused = false;
+    OwnableInitialize(owner);
+  }
+
+  function setup(
+    address _exchangeAddress,
+    address _prepaidCardManagerAddress,
+    address _provisioner,
+    address _tokenManager,
+    address[] calldata _trustedProvisioners,
+    address _versionManager
+  ) external onlyOwner {
+    exchangeAddress = _exchangeAddress;
+    prepaidCardManagerAddress = _prepaidCardManagerAddress;
+    provisioner = _provisioner;
+    tokenManager = _tokenManager;
+    versionManager = _versionManager;
+
+    for (uint256 i = 0; i < _trustedProvisioners.length; i++) {
+      _addTrustedProvisioner(_trustedProvisioners[i]);
+    }
+
+    emit Setup();
+  }
+
+  function setPaused(bool _paused) external onlyOwner {
+    paused = _paused;
+    emit PausedToggled(_paused);
+  }
+
+  function getTrustedProvisioners() external view returns (address[] memory) {
+    return trustedProvisioners.values();
+  }
+
+  function _addTrustedProvisioner(address _token) internal {
+    trustedProvisioners.add(_token);
+    emit TrustedProvisionerAdded(_token);
+  }
+
+  function removeTrustedProvisioner(address _token) external onlyOwner {
+    trustedProvisioners.remove(_token);
+    emit TrustedProvisionerRemoved(_token);
+  }
+
+  function provisionPrepaidCard(address customer, bytes32 sku)
+    external
+    returns (bool)
+  {
+    require(!paused, "Contract is paused");
+    require(
+      trustedProvisioners.contains(msg.sender),
+      "Only trusted provisioners allowed"
+    );
+    require(asks[sku] > 0, "Can't provision SKU with 0 askPrice");
```

Comment:
I think we are missing a require to ensure that there is enough balance for the issuer in order to create the prepaid card

Reply:
I agree, added this in https://github.com/cardstack/card-pay-protocol/pull/191/commits/bdb3e0d6f6766f8f9c1a914032cf0a57b3b8a808